django/db/models/base.py

```diff
@@ -1863,6 +1863,16 @@ def _check_long_column_names(cls, databases):
 
         return errors
 
+    @classmethod
+    def _get_check_or_condition_fields(cls, children):
+        fields = []
+        for child in children:
+            if isinstance(child, Q):
+                fields += cls._get_check_or_condition_fields(child.children)
+            elif isinstance(child[1], F):
```

Comment:
What if the rhs (`child[1]`) is not an F object, but another expression like`Lower('parent__name')`? Should we handle this as well?
Another thing is we don't check the left side of expression when the condition is like
` check=models.Q(parrent__age__lt=models.F('age'))`.

Reply:
Yes, this doesn't work for `Lower('parent__name')` I can change the patch to cover this case as well but I don't know exactly which kind of expression should I cover? and with more cases the code going to be complicated.
@felixxm Do you have any idea?

Reply:
If it's an expression its source expression tree should be walked (recursive `get_source_expressions`) and when the expression `isinstance(expr, str)` then you'd need to use split it using `LOOKUP_SEP`. The first part should be used to retrieve the field (`_meta.get_field(parts[0])`). If it's a related field (`field.remote_field is not None`) then you are trying to `JOIN` and it's disallowed.

Reply:
Something like

```python
elif hasattr(child[1], 'get_source_expressions'):
    for expr in child[1].get_source_expressions():
        if isinstance(expr, str):
           fields.add(expr.split(LOOKUP_SEP)[0])
        else:
            fields.extend(self._get_check_or_condition_fields(expr)
```

Reply:
Thanks @charettes 

I used your suggested code but added another if block:
```python
            elif hasattr(child[1], 'get_source_expressions'):
                for expr in child[1].get_source_expressions():
                    if isinstance(expr, str):  # couldn't find a usecase for this
                        fields.add(expr.split(LOOKUP_SEP)[0])
                    elif isinstance(expr, F):  # This is added by me
                        fields.add(expr.name)
                    else:
                        fields.update(cls._get_check_or_condition_fields(expr))
```
By this, we can cover [different kinds of checks](https://github.com/django/django/pull/12953/files#diff-bbed6332ed721c634e1b48b0eddd04a0R1427).

We still cannot check the left side of the check expression `check=models.Q(parrent__age__lt=10)`.
We can split the left side of the expression by `LOOKUP_SEP` and retrieve the field as you mentioned in your comment but this will add complexity to `self._get_check_or_condition_fields()`. Should I go for this?

Reply:
I think the function can be simplified to 


```python
@classmethod
def _get_expr_fields(cls, expr):
    fields = set()
    if isinstance(expr, Q):
       for child in expr.children:
           if isinstance(child, tuple):
               lookup, value = child
               fields.add(lookup.split(LOOKUP_SEP)[0])
               fields.add(cls._get_expr_fields(value))
           else:
               fields.update(cls._get_expr_fields(child[1]))
    elif isinstance(expr, F):
        fields.add(field.name)
    elif hasattr(expr, 'get_source_expressions'):
        for src_expr in expr.get_source_expression():
            if isinstance(src_expr, str):
                fields.add(src_expr.split(LOOKUP_SEP)[0])
            else:
                fields.update(cls._get_expr_fields(src_expr))
    return fields
```

Reply:
@charettes Thanks again for the new function.
But as I said, we will miss checking the left part of the expression. for example by this check:
`check=models.Q(parent__age=models.F('age')`

The `_get_expr_fields` function will find `fields={'age', 'parent'}` and when we call:
`cls._check_local_fields(fields, 'constraints')`
This won't return any error because `age` and `parent` are model fields.

Reply:
Is seems to me that we have to emit an error while traversing the condition. Even when join is on the right side, when fields are named the same, the `lookup.split(LOOKUP_SEP)` won't allow to reliably detect the violation `check=models.Q(age=models.F('parent__age')`. As @hramezani said, it will find `fields={'age', 'parent'}`. So we cannot use `_check_local_fields` approach and just pass the collected fields names. 
Second thought, do we actually allow `parent` field in the condition? If no, extending  `_check_local_fields` to forbid `ForeignKey` should do the trick. 

Another problem: `age__lte` condition on the left, split by `lookup.split(LOOKUP_SEP)` will yield `lte` field and fail the check. Any ideas on how to deal with lookups?

Reply:
@hramezani 

> The _get_expr_fields function will find fields={'age', 'parent'} and when we call:
cls._check_local_fields(fields, 'constraints')
This won't return any error because age and parent are model fields.

As @alextatarinov said you can't rely on `_check_local_fields` to determine whether or not a join would occur. You need to use a solution relying on `.remote_field is None` and `get_lookup`/`get_transform`.

That should do

```python
@classmethod
def _get_expr_references(cls, expr):
    references = set()
    if isinstance(expr, Q):
       for child in expr.children:
           if isinstance(child, tuple):
               lookup, value = child
               references.add(tuple(lookup.split(LOOKUP_SEP)))
               references.update(cls._get_expr_fields(value))
           else:
               references.update(cls._get_expr_fields(child[1]))
    elif isinstance(expr, F):
        references.add(tuple(field.name.split(LOOKUP_SEP)))
    elif hasattr(expr, 'get_source_expressions'):
        for src_expr in expr.get_source_expression():
            if isinstance(src_expr, str):
                references.add(tuple(src_expr.split(LOOKUP_SEP)))
            else:
                references.update(cls._get_expr_fields(src_expr))
    return references

references = self._get_expr_references(constraint.condition).union(constraint.check)
for field_name, *lookups in references:
    if not lookups:
        # If it has no lookups it cannot result in a JOIN.
        continue
    try:
        field = cls._meta.get_field(field_name)
    except FieldError:
        # handle missing field
    if not field.remote_field:
        # Only related fields can result in joins.
        continue
    # JOIN must happen at the first lookup.
    first_lookup = lookups[0]
    if field.get_transform(first_lookup) is None and field.get_lookup(first_lookup) is None:
        # This would result in a JOIN attempt, emit a warning
```

Reply:
What about m2m and reverse relationships? Something like `Q(m2m_cities=3)` will also produce the join.

Reply:
@alextatarinov we can add tests for it but from my understanding many-to-many field will be treated just like foreign key since `get_transform('cities') and get_lookup('cities')` will return `None`.

`_meta.get_field()` only return forward relationships so reverse one will result in a `FieldError`.

Reply:
@charettes Thanks for the clarification. Wasn't sure about `_meta.get_field()` behavior.

Reply:
Double-checked, `meta.get_field()` does actually work for m2m and reverse fields. Probably we need tests to verify. As I understand, the code above won't emit the warning for the m2m or reverse ForeignKey field.

```
User._meta.get_field('cities')
<django.db.models.fields.related.ManyToManyField: cities>

User._meta.get_field('profiles')
<ManyToOneRel: profile>
```

Reply:
Looks like we just need to use `_meta._get_fields(reverse=False)`.

Reply:
But it will still include m2m relations. I suggest something like this:
```
for field_name, *lookups in references:
    try:
        field = cls._meta.get_field(field_name)
    except FieldError:
        # handle missing field
    if not field.remote_field:
        # Only related fields can result in joins.
        continue
    if field.many_to_many or field.one_to_many:
        # Handle m2m / reverse ForeignKey
   
    if not lookups:
       # If it has no lookups it cannot result in a JOIN.
       continue
    # JOIN must happen at the first lookup.
    first_lookup = lookups[0]
    if field.get_transform(first_lookup) is None and field.get_lookup(first_lookup) is None
        # This would result in a JOIN attempt, emit a warning
```